Pull request

joaoalf opened the pull request:
Hi,

In my branch, I'm saving the request and response xml messages for debugging and for the UPS certification process. Also I added a field to pass the PackageServiceOptions/InsuredValue information and added the option to send a ShippingVoidRequest api call.

Please review.

Thanks.

Commit:
* Implement the Shipment Void action.
 * Save xml request and response messages.

Commit:
Fixed log issues

Commit:
Fixed view issues.

Commit:
Added function field msg_bin to return a binary version of msg.

Commit:
Fixed issue when voiding shipping without packages.

Commit:
Added field insured_value to the ups.shippingregister.package.

Commit:
Added insured_value field to the view.

sharoonthomas commented on shipping_register.py:
```diff
@@ -7,16 +7,22 @@
     ('confirmed', 'Confirmed'),
     ('accepted', 'Accepted'),
-    ('cancelled', 'Cancelled'),
+    ('canceled', 'Canceled'),
```
You cannot change this without writing a migration script.

This will break the code of every single openerp user using this module in production.

I dont see anything wrong with the spelling either according to http://www.merriam-webster.com/dictionary/cancel both spelling are valid. If you are particular about the way it appears, i suggest you use translations.

joaoalf replied:
It's true!! Both spellings are correct. I was just trying to make the spell more american friendly (http://www.reference.com/motif/language/cancelled-vs-canceled). I don't know why I did that, I saw the Cancelled button and near by in my screen I saw a Canceled button and I went to google to find out which one was the correct one and... I fliped a coin and the coin said “Canceled”!!

About break the code... Despite the previous existence of this state, there were no ways for an object to arrive in this state since there were no methods to cancel (Void) a shipment. It is as if we were creating this state right now. Am I wrong?

sharoonthomas replied:
Your argument about no way to reach the state is true, since there was no void facility. However, i think its always good to write backward combatible code, lets assume the customer had a customisation on a different module etc.

sharoonthomas commented on shipping_register.py:
```diff
@@ -44,6 +50,32 @@ class UpsCodes(osv.osv):
+
+    _name = 'ups.message'
+
+    def _text2bin(self, cr, uid, ids, field_name, arg, context):
```
Please use PEP008 styling and use cursor, user, .. like other functions/methods in this module

sharoonthomas commented on shipping_register.py:
```diff
@@ -44,6 +50,32 @@ class UpsCodes(osv.osv):
+    Model to record all XML messages exchanged between the client and the server.
+    """
+
+    _name = 'ups.message'
```
Please add a _description too

sharoonthomas commented on shipping_register.py:
```diff
@@ -7,16 +7,22 @@
+    ('canceled', 'Canceled'),
 ]
 
+MESSAGE_TYPE_SELECTION = [
```
Please define this in the selection field itself because its used in only _one_ place

sharoonthomas commented on shipping_register.py:
```diff
@@ -44,6 +50,32 @@ class UpsCodes(osv.osv):
+                                 type='selection',
+                                 readonly=True),
+        'msg': fields.text('XML Message', readonly=True),
+        'msg_bin': fields.function(_text2bin, string='XML Message', type='binary', method=True, readonly=True),
```
the line is too long.

Try to wrap to 80 chars (PEP008)

Again, I dont see why this field needs to be there. Even if there is XML, openerp will encode it in boh XML-RPC and other protocols before displaying. If you are facing any issue there, it could be a protocol handling error on openerp server

joaoalf replied:
I did this 'trick' to be able to see the msg field as text and at the same time be able to download and save it's content.

sharoonthomas commented on shipping_register.py:
```diff
@@ -44,6 +50,32 @@ class UpsCodes(osv.osv):
+
+    _columns = {
+        'name': fields.char('Message name', size=50, readonly=True),
+        'type': fields.selection(MESSAGE_TYPE_SELECTION,
```
the lines are not properly aligned acc to PEP008

sharoonthomas commented on shipping_register.py:
```diff
@@ -44,6 +50,32 @@ class UpsCodes(osv.osv):
+                                 string='Message type',
+                                 type='selection',
+                                 readonly=True),
+        'msg': fields.text('XML Message', readonly=True),
```
Please use `message` as the field name too.

sharoonthomas commented on shipping_register.py:
```diff
@@ -104,6 +139,8 @@ def _add_packages(self, cursor, user, register_id, context=None):
                 Weight=str(package.weight), Code=ups_uoms[0], 
                 Description=package.description or 'None')
+            package_service_options = ShipmentConfirm.package_service_options_type(
+                ShipmentConfirm.insured_value_type(MonetaryValue=str(package.insured_value) or ''))
```
Try to avoid crossing 80 chars

sharoonthomas commented on shipping_register.py:
```diff
@@ -44,6 +50,32 @@ class UpsCodes(osv.osv):
+        return res
+
+    _columns = {
+        'name': fields.char('Message name', size=50, readonly=True),
```
Looking at your code and what fills this field, i think name is a misnomer. It should probably be

sharoonthomas commented on shipping_register.py:
```diff
@@ -79,9 +111,12 @@ def get_ups_api(self, cursor, user, call='confirm', context=None):
+        elif call == 'accept':
             return ShipmentAccept(ups_credentials[0], 
                 ups_credentials[1], ups_credentials[2], ups_credentials[3])
+        else:
```
To avoid poor api design like what I had done with the if else, you should explicitly check if the the call is 'void' to return this api.

So that the next time some other method is implemented, this is not broken.

sharoonthomas commented on shipping_register.py:
```diff
@@ -334,9 +404,89 @@ def accept_price(self, cursor, user, ids, context=None):
+                'void', context)
+                
+            try:
+                #response = shipment_accept_instance.request(shipment_accept)
```
PLease remove commented out code

sharoonthomas commented on shipping_register.py:
```diff
@@ -334,9 +404,89 @@ def accept_price(self, cursor, user, ids, context=None):
+                
+            try:
+                #response = shipment_accept_instance.request(shipment_accept)
+                request, response = shipment_void_instance.request(shipment_void, return_request=True)
```
more than 80 chars

sharoonthomas commented on shipping_register.py:
```diff
@@ -244,9 +283,20 @@ def do_shipping_request(self, cursor, user, ids, context=None):
+                                    'msg': error[1]}),
+                            (0, 0, {'name': 'ConfirmShipmentResponse',
+                                    'type': 'response',
+                                    'msg': etree.tostring(error[2], pretty_print=True)})]
```
This is not guaranteed to work. The error trapped here and the exception handled here is very generic `Exception`

If for example a connectivity error or some other error occurs, it will not have the two parts you are expecting.

If you are sure you want to store these messages, then handle `PyUPSException`

sharoonthomas commented on shipping_register.py:
```diff
@@ -312,9 +370,21 @@ def accept_price(self, cursor, user, ids, context=None):
+                                    'msg': error[1]}),
+                            (0, 0, {'name': 'AcceptShipmentResponse',
+                                    'type': 'response',
+                                    'msg': etree.tostring(error[2], pretty_print=True)})]
```
as i said before this is not guaranteed to work

sharoonthomas commented on shipping_register.py:
```diff
@@ -334,9 +404,89 @@ def accept_price(self, cursor, user, ids, context=None):
+                                    'msg': error[1]}),
+                            (0, 0, {'name': 'VoidShipmentResponse',
+                                    'type': 'response',
+                                    'msg': etree.tostring(error[2], pretty_print=True)})]
```
as i said before, this is not guaranteed to work

sharoonthomas commented on shipping_register.py:
```diff
@@ -334,9 +404,89 @@ def accept_price(self, cursor, user, ids, context=None):
+                                'type': 'response',
+                                'msg': etree.tostring(response, pretty_print=True)})]
+            
+            self.write(cursor, user, shipping_register_record.id, 
```
please use PEP 008 styling on code

sharoonthomas commented on shipping_register.py:
```diff
@@ -355,6 +505,9 @@ def accept_price(self, cursor, user, ids, context=None):
         'saturday_delivery': fields.boolean('Saturday Delivery?'),
         'description': fields.text('Description'),
         'state':fields.selection(STATE_SELECTION, 'Status', readonly=True,),
+        'xml_msgs': fields.one2many('ups.message',
```
use pep008 styling on code

sharoonthomas commented on shipping_register_view.xml:
```diff
@@ -52,6 +55,35 @@
+                    <field name="name" />
+                    <field name="type" />
+                    <field name="shipping_register_rel" />
+		    <newline />
```
the code isnt aligned, you might be mixing tabs and spaces

sharoonthomas commented on shipping_register_view.xml:
```diff
@@ -85,18 +117,23 @@
                                 <field name="billed_weight_uom" />
                                 <field name="total_amount" />
                                 <field name="total_amount_currency" />
+				<field name="xml_msgs" widget="one2many_list" nolabel="1" colspan="4" />
```
the code isnt aligned

sharoonthomas commented on shipping_register_view.xml:
```diff
@@ -85,18 +117,23 @@
                         </notebook>
-                        <field name="state" />
+                        <field name="state" colspan="4"/>
+			<newline />
```
the code isnt aligned

sharoonthomas commented on shipping_register_view.xml:
```diff
@@ -52,6 +55,35 @@
+            <field name="model">ups.message</field>
+            <field name="type">form</field>
+            <field name="arch" type="xml">
+                <tree string="XML Messages">
```
should be form ?

sharoonthomas commented:
Hi,

Sorry for the comments, but we maintain a high quality standard for openlabs modules unlike most of the openerp modules out there.

I strongly suggest that you use codereview.appspot.com to send code for review before you send a pull request so that this can be ensured.

Remember to add your copyrights too, or may be i should add a CREDITS file.

If you need help on pushing to codereview, see : http://code.google.com/p/tryton/wiki/HowtoContribute

shalabhaggarwal commented on shipping_register.py:
```diff
@@ -44,6 +50,32 @@ class UpsCodes(osv.osv):
+                                 readonly=True),
+        'msg': fields.text('XML Message', readonly=True),
+        'msg_bin': fields.function(_text2bin, string='XML Message', type='binary', method=True, readonly=True),
+        'shipping_register_rel': fields.many2one('ups.shippingregister', 'Relation Field', readonly=True)
```
Try to make field names as what they actually do.
The below seems more readable.

'shipping_register: fields.many2one('ups.shippingregister', 'Shipping Register', readonly=True)

Commit:
* PEP8
* Fixed alignment issue on the view code.

Commit:
Pre-merge adjustments

- Changed version to 1.0.1
- Changed field names (msg, msg_bin and xml_msgs) to (message,
  message_bin and xml_messages)
- PEP8fied __openerp__.py

Commit:
Reverted the 'canceled' spelling to 'cancelled'

Commit:
Added the 'Save XML ?' to the company configuration.

* Code clean-up
* Some backward compatibility with previous PyUPS versions.

max3903 commented:
2012-08-09 21:45:53,328 42216 ERROR ***************\* openerp.sql_db: bad query: INSERT INTO ir_model (id,model, name, info,state) VALUES (435, E'ups.message', E'Model to record all XML messages exchanged between the client and
    the server.
    ', E'Model to record all XML messages exchanged between the client and
    the server.
    ', E'base')
Traceback (most recent call last):
  File "/Users/openerp/openerp-virt/lib/python2.6/site-packages/openerp-6.1-py2.6.egg/openerp/sql_db.py", line 212, in execute
    res = self._obj.execute(query, params)
DataError: value too long for type character varying(64)

Commit:
Shortened ups.message's description string.

max3903 commented:
Any reason why this is not merged ?

sharoonthomas commented:
Never received an alert for this one. 
Will review ASAP, thanks for the reminder.

sharoonthomas commented:
Thanks for the pull request. I have squashed your commits into develop at 5a20d795989f48438f9e2a650da51bad69adcd09. The new version 1.1.0 includes all your changes.

Thanks again for your contribution and apology for the delay.
